### ISSUE TITLE
Expose `Node.toXmlNode()` as public API

### ISSUE THREAD

**Issue by SimonMarquis** [martinbonnin/xoxo#12]
In Android Lint `ResourceXmlDetector`, when parsing strings resources, we usually declare this:

```kotlin
override fun getApplicableElements() = listOf(TAG_STRING, TAG_PLURALS, TAG_STRING_ARRAY)
```

which triggers this callback on each elements:

```kotliln
override fun visitElement(context: XmlContext, element: Element)
```

We only have access to the `Element`, and unfortunately neither the `Node.toXmlNode()` extension or the `XmlElement` constructor are public. Which makes it impossible to use (without internal code suppression annotation 🙈)

Would you be ok to open one of these? I suppose exposing the extension makes more sense here, unless it was private for another reason than limiting the API surface.

**Comment by martinbonnin** [comment]
> Would you be ok to open one of these? I suppose exposing the extension makes more sense here, unless it was private for another reason than limiting the API surface.

Limiting the API surface was the reason. Always easier to add than remove. But makes sense 👍 . I'll merge and try to make a release this week.

**Comment by SimonMarquis** [comment]
@martinbonnin there is also one thing that is generally needed when dealing with Lint reports, which is to provide the location/scope.

```kotlin
context.report(
    /*...*/
    location = context.getLocation(node),
    /*...*/
)
```

unfortunately, xoxo's `XmlNode` implementations do not expose them explicitly, which makes it impossible to use the appropriate scope (we only have access to the element from `fun visitElement(context: XmlContext, element: Element)`).

I was wondering what would be the best solution here (both of them open-up even more the API surface 😖):
- make `XmlNode` implement the `org.w3c.dom.Node` interface (this creates a collision with `textContent` getter)
- expose the property delegates in `XmlText` and `XmlElement` (they are currently private)

**Comment by martinbonnin** [comment]
@SimonMarquis can you share a sample XML you'd like to parse?

**Comment by SimonMarquis** [comment]
I feel like I'm trying to use the library for something that was not the main purpose, so feel free to tell me.

Here for example a basic Android strings file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <string name="string">...</string>
    <plurals name="plurals">
        <item quantity="other">...</item>
    </plurals>
    <string-array name="string-array">
        <item>...</item>
    </string-array>
</resources>
```

I'm using this basic Lint detector:

```kotlin
class MyDetector : ResourceXmlDetector() {
    override fun appliesTo(folderType: ResourceFolderType) = folderType == VALUES
    override fun getApplicableElements(): Collection<String> = listOf(TAG_STRING, TAG_PLURALS, TAG_STRING_ARRAY)

    override fun visitElement(context: XmlContext, element: Element) {
        val element = XmlElement(element)
        when (element.name) {
            TAG_STRING -> element.walk()
                .filterIsInstance<XmlText>()
                .forEach { checkText(...) }
            TAG_PLURALS, TAG_STRING_ARRAY -> element.childElements
                .filter { it.name == TAG_ITEM }
                .flatMap { it.walk().filterIsInstance<XmlText>() }
                .forEach { checkText(...) }
        }
    }

    private fun checkText(context: XmlContext, node: XmlText) {
        // ...
        context.report(
            // ...
            scope = node/*.text*/,
            location = context.getLocation(node/*.text*/),
            // ...
        )
    }

}
```

To report the lint issue at the right place, I need to track back to the `org.w3c.dom.Element`/`org.w3c.dom.Node`.

**Comment by martinbonnin** [comment]
I see. Thanks for providing this 🙏 

This only works in the Android lint context because they're parsing the document using their own parser and [setting a tag on the element](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:common/src/main/java/com/android/utils/PositionXmlParser.java;l=932?q=PositionXmlPar):

```java
       public void startElement(String uri, String localName, String qName,
                Attributes attributes) throws SAXException {
                // ...
                Position pos = getCurrentPosition();
                
                // ...
                element.setUserData(POS_KEY, findOpeningTag(pos), null);
```

What about adding `XmlText.asW3CText()` and `XmlElement.asW3CElement()`. Same pattern as `kotlinx.DateTime.asJavaDateTime()`? That sounds reasonnable

It could even be `XmlNode.asW3CNode()`:

```kotlin
sealed interface XmlNode {
    fun asW3CNode(): Node
}
```

With covariant overrides for `XmlText` and `XmlElement`:

```kotlin
class XmlElement internal constructor(private val element: Element) : XmlNode {
  override fun asW3CNode(): Element = element
}
```

edit: Or maybe `asDomNode()` would read a bit better?

**Comment by SimonMarquis** [comment]
I agree! `fun asDomNode(): Node` on the sealed interface would solve my need 👍

**Comment by martinbonnin** [comment]
Neat! Do you want to open the PR?

**Comment by SimonMarquis** [comment]
Yes, let me quickly address this 💨
